aptos-move/mvhashmap/src/versioned_code.rs

```diff
@@ -0,0 +1,214 @@
+    /// blob or CompiledModule, but must satisfy TransactionWrite to be able to
+    /// generate the hash below.
+    module: Arc<V>,
+    /// The hash of the blob, used instead of incarnation for validation purposes,
+    /// and also for uniquely identifying associated executables. None implies
+    /// that the stored module (WriteOp) actually corresponded to a deletion.
+    maybe_hash: Option<HashValue>,
+}
+
+/// A VersionedValue internally contains a BTreeMap from indices of transactions
+/// that update the given access path alongside the corresponding entries.
+struct VersionedValue<V: TransactionWrite, X: Executable> {
+    versioned_map: BTreeMap<TxnIndex, CachePadded<Entry<V>>>,
+
+    /// Executable based on the storage version of the module.
+    base_executable: Option<Arc<X>>,
+    /// Executables corresponding to published versions of the module, based on hash.
+    executables: HashMap<HashValue, Arc<X>>,
+}
+
+/// Maps each key (access path) to an interal VersionedValue.
+pub struct VersionedCode<K, V: TransactionWrite, X: Executable> {
+    values: DashMap<K, VersionedValue<V, X>>,
+}
+
+impl<V: TransactionWrite> Entry<V> {
+    pub fn new_write_from(flag: usize, module: V) -> Entry<V> {
+        let maybe_hash = module.extract_raw_bytes().map(|bytes| {
+            let mut hasher = DefaultHasher::new(b"Module");
+            hasher.update(&bytes);
+            hasher.finish()
+        });
+
+        Entry {
+            flag: AtomicUsize::new(flag),
+            module: Arc::new(module),
+            maybe_hash,
+        }
+    }
+
+    pub fn flag(&self) -> usize {
+        self.flag.load(Ordering::Acquire)
+    }
+
+    pub fn mark_estimate(&self) {
+        self.flag.store(FLAG_ESTIMATE, Ordering::Release);
+    }
+}
+
+impl<V: TransactionWrite, X: Executable> VersionedValue<V, X> {
+    pub fn new() -> Self {
+        Self {
+            versioned_map: BTreeMap::new(),
+            base_executable: None,
+            executables: HashMap::new(),
+        }
+    }
+}
+
+impl<V: TransactionWrite, X: Executable> Default for VersionedValue<V, X> {
+    fn default() -> Self {
+        VersionedValue::new()
+    }
+}
+
+impl<K: Hash + Clone + Eq, V: TransactionWrite, X: Executable> VersionedCode<K, V, X> {
+    pub(crate) fn new() -> Self {
+        Self {
+            values: DashMap::new(),
+        }
+    }
+
+    pub(crate) fn mark_estimate(&self, key: &K, txn_idx: TxnIndex) {
+        let v = self.values.get(key).expect("Path must exist");
+        v.versioned_map
+            .get(&txn_idx)
+            .expect("Entry by the txn must exist to mark estimate")
+            .mark_estimate();
+    }
+
+    pub(crate) fn delete(&self, key: &K, txn_idx: TxnIndex) {
+        // TODO: investigate logical deletion.
+        let mut v = self.values.get_mut(key).expect("Path must exist");
+        v.versioned_map.remove(&txn_idx);
+    }
+
+    fn read(
```

Comment:
"read_moudle" to be consistent?

Reply:
not sure I follow anymore, it's a private fn.
consistent with what?

Reply:
read_data and read_code now